torch/csrc/jit/script/module.h

```diff
@@ -105,124 +122,133 @@ struct NamedParameter {
     }
+    values_.push_back(std::move(value));
+    index_[name] = values_.size() - 1;
+    return values_.back();
+  }
+  at::optional<T&> find(const std::string& str) {
+    auto it = index_.find(str);
+    if(it == index_.end())
+      return at::nullopt;
+    return at::optional<T&>(values_.at(it->second));
+  }
+  at::optional<const T&> find(const std::string& str) const {
+    auto it = index_.find(str);
+    if(it == index_.end())
+      return at::nullopt;
+    return at::optional<const T&>(values_.at(it->second));
+  }
+  T& get(const std::string& name, const char * what) {
+    if(auto v = find(name)) {
+      return *v;
+    }
+    std::stringstream ss;
+    ss << "module " << what << "'" << name << "' is not defined.";
+    throw std::runtime_error(ss.str());
+  }
+  const T& get(const std::string& name, const char * what) const {
+    if(auto v = find(name)) {
+      return *v;
+    }
+    std::stringstream ss;
+    ss << "module " << what << "'" << name << "' is not defined.";
+    throw std::runtime_error(ss.str());
+  }
+  const std::vector<T>& values() const {
+    return values_;
   }
+private:
+  std::unordered_map<std::string, size_t> index_;
+  std::vector<T> values_;
 };
 
 struct Module : public std::enable_shared_from_this<Module> {
   TH_DISALLOW_COPY_AND_ASSIGN(Module);
   Module(bool optimize)
   : optimize(optimize) {}
 
-  void register_parameter(const std::string & name, at::Tensor v) {
-    parameters.push_back(NamedParameter(name, std::move(v)));
-    add_member(name, NamedMember::Parameter, parameters.size() - 1);
-  }
-  void register_or_set_parameter(const std::string & name, autograd::Variable v) {
-    if(find_attribute(name) == NamedMember::Parameter) {
-      set_parameter(name, v);
-    } else {
-      register_parameter(name, v);
+  void register_parameter(const std::string & name, autograd::Variable v, bool is_buffer) {
+    if(auto p = parameters.find(name)){
+      *p->slot() = v;
+      p->is_buffer = is_buffer;
+      return;
     }
+    parameters.insert(name, NamedParameter(name, std::move(v), is_buffer), "parameter");
   }
   void register_module(const std::string& name, std::shared_ptr<Module> module) {
-    JIT_ASSERT(module);
-    modules.push_back(NamedModule {name, std::move(module)});
-    add_member(name, NamedMember::Module, modules.size() - 1);
+    modules.insert(name, {name, std::move(module)}, "module");
   }
 
-  Method& create_method(const std::string & name) {
-    methods.emplace_back(new Method(name, optimize));
-    add_member(name, NamedMember::Method, methods.size() - 1);
-    return *methods.back();
+  Method& create_method(const std::string & name, std::shared_ptr<Graph> graph = nullptr, std::vector<at::Tensor*> member_inputs = {}) {
+    if(!graph)
+      graph = std::make_shared<Graph>();
+    std::unique_ptr<Method> method(new Method(name, optimize, std::move(graph), std::move(member_inputs)));
+
+    return *methods.insert(name, std::move(method), "method");
```

Comment:
We don't need an *ordered* dict for methods right?

Reply:
Also, it looks like the `what` arguments are always the same for any given dict. Can't we make them a value given to the constructor, so we can avoid passing them in every time?